### PR TITLE
Improve JSDoc comments

### LINKED ISSUE #1106: Update JSDoc
In trying to better understand how Preact works, I've noticed that the JSDocs of some of the internal methods [have become out of date](https://github.com/developit/preact/blob/4f57ba02bd4dee37d3278f2e75ce18073a36deac/src/vdom/component.js#L13). I'd like to take a stab at updating these in an effort to better understand Preact. But before I do, could I get other's opinions on what style they prefer? I see a couple different styles in the codebase today. Some options are outlined below.

**Comment start**
*Most common in Preact*
Comment starts on same line as `/**`. ([example in Preact](https://github.com/developit/preact/blob/4f57ba02bd4dee37d3278f2e75ce18073a36deac/src/vdom/component.js#L10)):
```js
/** Description
 *	@param {Object} props Description
 */
```

Comment starts on line after `/**` ([example in Preact](https://github.com/developit/preact/blob/4f57ba02bd4dee37d3278f2e75ce18073a36deac/src/util.js#L2)) (Note: this appears to be the style on [usejsdoc.org](http://usejsdoc.org/about-getting-started.html))
```js
/**
 * Description
 * @param {Object} props Description
 */
```

With both of these, the intention is that `@tag` aligns to the top `Description`.

**Parameter description alignment**
Tabs between `@tag {type} name` and `description` to align descriptions ([example in Preact](https://github.com/developit/preact/blob/4f57ba02bd4dee37d3278f2e75ce18073a36deac/src/vdom/component.js#L13)):
```js
/**
 * Description
 * @param {Object} props	Description
 * @param {Number} props2	Description2
 */
```

Single space between `@tag {type} name` (i.e. don't align descriptions) ([example in Preact](https://github.com/developit/preact/blob/4f57ba02bd4dee37d3278f2e75ce18073a36deac/src/vdom/index.js#L40))
```js
/**
 * Description
 * @param {Object} props Description
 * @param {Number} props2 Description2
 */
```

@developit @marvinhagemeister - let me know if you have an opinion or not - it doesn't matter to me. I just wanted to see if it matters to you.

### PR DESCRIPTION
**Pull request by andrewiggins**
Update Preact's JSDoc comments to better reflect the current code and improve editor experience (if using editor with TypeScript support/extension). I make use of `@typedef`s to define a couple of key complex types used by Preact. Each complex type is described below.

## Complex Types

**Options**: Documents all the public options a consumer of Preact can define

**VNode**: Documents the properties of a Preact Virtual DOM Node

**PreactElement**: Documents all of the properties Preact adds to the DOM elements it creates

## What about Component?

I haven't exhaustively documented Component in this PR because that is a much more complex type to document in JSDoc and I'm considering doing it in its own PR to discuss different ways of how to accomplish it. I've started experimenting with how to do this in my [update-jsdoc-component-doc branch](https://github.com/andrewiggins/preact/blob/update-jsdoc-component-doc/src/component.js) if you are curious.

## Other Notes

These JSDoc comments rely on a new feature in Typescript 2.9, "[allow `import(...)-ing` types at any location](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#allow-import-ing-types-at-any-location)". I use this feature to import `@typedef`s across files. The next version of VSCode should include TS 2.9 by default so these comments will work by default in that version. If you'd like to check out the TS 2.9 experience now, check out this branch and set the VSCode workspace setting `typescript.tsdk` to `node_modules\\typescript\\lib`.

This PR does not eliminate all errors that appear in the editor. In order to do that Preact would likely need to embrace full TypeScript or Flow syntaxes to utilize the entirety of features those languages offer. It might be possible to remove all errors using only JSDoc syntax however I'd like to explore that possibility in a separate PR as it might require an additional significant amount of changes.

I've rewritten many JSDoc comments to the default style that VSCode uses when you add new JSDoc comments.

Fixes #1106

### CODE REVIEW

File: src/options.js [hunk 1]
```diff
@@ -1,27 +1,22 @@
...
-	 */
-	//vnode(vnode) { }
-
-	/** Hook invoked after a component is mounted. */
-	// afterMount(component) { }
+/**
+ * Global options
+ * @public
+ * @typedef Options
+ * @property {Boolean} [syncComponentUpdates] If `true`, `prop` changes trigger synchronous component updates. Defaults to true.
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/options.js]: Yup, good call. Missed this one.

File: src/util.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 /**
- *  Copy all properties from `props` onto `obj`.
- *  @param {Object} obj		Object onto which properties should be copied.
- *  @param {Object} props	Object from which to copy properties.
- *  @returns obj
- *  @private
+ * Copy all properties from `props` onto `obj`.
+ * @param {object} obj Object onto which properties should be copied.
+ * @param {object} props Object from which to copy properties.
+ * @returns {object} obj
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/util.js]: Removed

File: src/h.js [hunk 3]
```diff
@@ -28,9 +28,9 @@ const EMPTY_CHILDREN = [];
  *
  * `h('div', { id: 'foo', name : 'bar' }, 'Hello!');`
  *
- * @param {string} nodeName	An element name. Ex: `div`, `a`, `span`, etc.
- * @param {Object} attributes	Any attributes/props to set on the created element.
- * @param rest			Additional arguments are taken to be children to append. Can be infinitely nested Arrays.
+ * @param {string | Function} nodeName An element name. Ex: `div`, `a`, `span`, etc.
+ * @param {object} attributes Any attributes/props to set on the created element.
+ * @param {VNode[]} [rest] Additional arguments are taken to be children to append. Can be infinitely nested Arrays.
```
[the remark this answers is not in the document]
Comment (marvinhagemeister) [reply src/h.js]: oh my bad, totally missed the square brackets 🙈

**Comment by marvinhagemeister** [comment]
@andrewiggins Thanks for addressing the comments. Hope you don't mind that I was rather nitpicky about the types. Only thing left is wether we use jsdoc or full on TS types syntax. I pinged everyone on slack and we should have an answer soon 👍

File: src/dom/index.js [hunk 4]
```diff
@@ -1,36 +1,63 @@
...
 import options from '../options';
 
+/**
+ * A DOM event listener
+ * @typedef {(e: Event) => void} EventListner
+ */
+
+/**
+ * A mapping of event types to event listeners
+ * @typedef {{ [eventType: string]: EventListener}} EventListenerMap
```
[the remark this answers is not in the document]
Comment (developit) [reply src/dom/index.js]: I'm ambivalent here. We might find the full-on TS syntax a bit heavy, which could hurt contribution. But, it likely provides better docs (which I think is our end goal here).

File: src/dom/index.js [hunk 5]
```diff
@@ -1,36 +1,63 @@
...
- *	@param {any} old	The last value that was set for this name/node pair
- *	@param {any} value	An attribute value, such as a function to be used as an event handler
- *	@param {Boolean} isSvg	Are we currently diffing inside an svg?
- *	@private
+/**
+ * Set a named attribute on the given Node, with special behavior for some names and event handlers.
+ * If `value` is `null`, the attribute/handler will be removed.
+ * @param {PreactElement} node An element to mutate
+ * @param {string} name The name/key to set, such as an event or attribute name
+ * @param {any} old The last value that was set for this name/node pair
```
[the remark this answers is not in the document]
Comment (developit) [reply src/dom/index.js]: Depends on what we're using the comments for. If we're generating docs from them and `*` doesn't give the necessary output, we kinda have to go with `any`.  If either work, my preference would be slightly in favor of the JSDoc-compatible types.

File: src/clone-element.js [hunk 6]
```diff
@@ -3,9 +3,9 @@ import { h } from './h';
 
 /**
  * Clones the given VNode, optionally adding attributes/props and replacing its children.
- * @param {VNode} vnode		The virtual DOM element to clone
- * @param {Object} props	Attributes/props to add when cloning
- * @param {VNode} rest		Any additional arguments will be used as replacement children.
+ * @param {import('./vnode').VNode} vnode The virtual DOM element to clone
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/clone-element.js]: Oh true! I'll do that.

File: src/constants.js [hunk 7]
```diff
@@ -1,13 +1,17 @@
 // render modes
 
+/** Do not recursively re-render a component */
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/constants.js]: Yea that's right, I believe

File: src/dom/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/dom/index.js]: Sounds like we are learning towards JSDoc syntax so I've updated the types to use JSDoc type syntax where possible (though I could use an extra set of eyes to make sure I didn't miss anything).

3 places where I think we should (or have to in some cases) use TS syntax:
1. (should) Use TS function syntax (e.g. `(arg: string) => string`) in some places so we can specify the parameter types
2. (must) Use TS type operations (namely `&` and `|`) because I don't know of a JSDoc equivalent
3. (must) Use TS 2.9 import type syntax (e.g. `import('./vnode').VNode`) because I don't know of a JSDoc equivalent

File: src/dom/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/dom/index.js]: Per my reply on the other comment, I've updated this to `*`. TS correctly parses and understands the `*`.

File: src/dom/index.js [hunk 8]
```diff
@@ -91,8 +118,12 @@ export function setAccessor(node, name, old, value, isSvg) {
...
 
 
-/** Attempt to set a DOM property to the given value.
- *	IE & FF throw for certain property-value combinations.
+/**
+ * Attempt to set a DOM property to the given value.
+ * IE & FF throw for certain property-value combinations.
+ * @param {Node} node The node to set a property on
+ * @param {string} name The name of the property to set
+ * @param {any} value The value to set
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/dom/index.js]: Updated to `*`

File: src/h.js [hunk 9]
```diff
@@ -28,9 +28,9 @@ const EMPTY_CHILDREN = [];
  *
  * `h('div', { id: 'foo', name : 'bar' }, 'Hello!');`
  *
- * @param {string} nodeName	An element name. Ex: `div`, `a`, `span`, etc.
- * @param {Object} attributes	Any attributes/props to set on the created element.
- * @param rest			Additional arguments are taken to be children to append. Can be infinitely nested Arrays.
+ * @param {string | Function} nodeName An element name. Ex: `div`, `a`, `span`, etc.
+ * @param {object} attributes Any attributes/props to set on the created element.
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/h.js]: Add `null` to possible types

File: src/vdom/component-recycler.js [hunk 10]
```diff
@@ -1,20 +1,31 @@
 import { Component } from '../component';
 
-/** Retains a pool of Components for re-use, keyed on component name.
- *	Note: since component names are not unique or even necessarily available, these are primarily a form of sharding.
- *	@private
+/**
+ * Retains a pool of Components for re-use, keyed on component name.
+ * Note: since component names are not unique or even necessarily available, these are primarily a form of sharding.
+ * @type {Record<string, Component[]>}
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/vdom/component-recycler.js]: Updated to `Object.<string, Component[]>`

File: src/vdom/component.js [hunk 11]
```diff
@@ -7,11 +7,13 @@ import { diff, mounts, diffLevel, flushMounts, recollectNodeTree, removeChildren
...
-/** Set a component's `props` (generally derived from JSX attributes).
- *	@param {Object} props
- *	@param {Object} [opts]
- *	@param {boolean} [opts.renderSync=false]	If `true` and {@link options.syncComponentUpdates} is `true`, triggers synchronous rendering.
- *	@param {boolean} [opts.render=true]			If `false`, no render will be triggered.
+/**
+ * Set a component's `props` and possibly re-render the component
+ * @param {import('../component').Component} component The Component to set props on
+ * @param {object} props The new props
+ * @param {number} opts Render options - specifies how to re-render the component
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/vdom/component.js]: Yea, agreed. Renamed to `renderMode`.

File: src/vdom/component.js [hunk 12]
```diff
@@ -53,11 +55,13 @@ export function setComponentProps(component, props, opts, context, mountAll) {
...
 
-/** Render a Component, triggering necessary lifecycle events and taking High-Order Components into account.
- *	@param {Component} component
- *	@param {Object} [opts]
- *	@param {boolean} [opts.build=false]		If `true`, component will build and store a DOM node if not already associated with one.
- *	@private
+/**
+ * Render a Component, triggering necessary lifecycle events and taking High-Order Components into account.
+ * @param {import('../component').Component} component The component to render
+ * @param {number} [opts] A number indicating the kind of render to execute (e.g. sync, async, force)
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/vdom/component.js]: Renamed to renderMode

File: src/vnode.js [hunk 13]
```diff
@@ -1,2 +1,9 @@
-/** Virtual DOM Node */
+/**
+ * Virtual DOM Node
+ * @typedef VNode
+ * @property {string | Function} nodeName The string of the DOM node to create or Component constructor to render
+ * @property {Array<VNode | string>} children The children of node
+ * @property {string | number | undefined} key The key used to identify this VNode in a list
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/vnode.js]: Technically the types mean two different things when used as function return types in the TS type-checker (at least for now).

For example ([TS playground](http://www.typescriptlang.org/play/index.html#src=function%20test()%3A%20void%20%7B%20%7D)),
```ts
function (): void {}
```
The above code uses `void`, instructing TypeScript that the the function does not need to have a return statement.

However, in [this next example](http://www.typescriptlang.org/play/index.html#src=function%20test()%3A%20undefined%20%7B%20%7D),
```ts
function (): undefined { return undefined; }
```
instructs TypeScript that all code paths in the function should lead to a return statement that returns `undefined`.

So in summary, I suggest use undefined, unless you want to specify that a function doesn't have to have a return statement, then use `void`. For Preact, the default implementation is just `render() {}`, so `void` is the right type to use there. Everywhere else should use `undefined`.

File: src/clone-element.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/clone-element.js]: FYI, Had to add VNode to the eslint no-unused-var exception

File: src/component.js [hunk 14]
```diff
@@ -3,52 +3,48 @@ import { extend } from './util';
...
+	 * @public
+	 * @type {object}
 	 */
 	this.state = this.state || {};
 }
 
 
 extend(Component.prototype, {
 
-	/** Returns a `boolean` indicating if the component should re-render when receiving the given `props` and `state`.
```
[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/component.js]: Yea, I agree - documenting all the lifecycles here in JSDoc without having to define them on the prototype would be ideal. I'm experimenting with some techniques in a branch in my fork ([a commmit with one experiement](https://github.com/andrewiggins/preact/commit/f1ad967b4f6df8acbedc329c812ac6bbbbb4e1a4)), but so far they all have some compromises. If I get to a good place with a good approach I'll open a PR.

**Comment by andrewiggins** [comment]
@marvinhagemeister @developit I believed I've resolved all the comments on this PR. When you have time, let me know what you think. Thanks for taking the time to leave your thoughts.

@marvinhagemeister No worries about the "nitpicky"-ness :smile: I thought your comments were were helpful.

File: src/clone-element.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (marvinhagemeister) [reply src/clone-element.js]: I feel like we should be consistent with this where possible. The latest changes have a top level `import` here and all other places use `import` inside jsdoc.

[the remark this answers is not in the document]
Comment (marvinhagemeister) [reply src/clone-element.js]: Although I think the top level imports are easier to understand for new contributors, I'm a bit worried about reducing usefulness of `no-unused-var-exception`. After thinking a bit more about it I think I'll actually prefer the inline imports. What are your thoughts about this? @developit @andrewiggins

[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/clone-element.js]: Yea, that's fair. Consistency would be good. I don't have a strong preference for either pattern but since Preact is not TypeScript, I like the clean separation of type-related "code" (i.e. JSDoc) and real code. Putting the unused import at the top of the file may seem like a bug to new contributors since no real code uses it.

[the remark this answers is not in the document]
Comment (andrewiggins) [reply src/clone-element.js]: @marvinhagemeister I've gone ahead and reverted the change back to importing the type inside the JSDoc comment to be consistent with the rest of the changes in this PR. I suggest if we want to use ES6 imports purely for JSDoc comments, we can do that in another PR that applies it consistently across the codebase. Doing that we can evaluate what it does to the codebase and better consider the maintenance tradeoffs, etc.

[the remark this answers is not in the document]
Comment (marvinhagemeister) [reply src/clone-element.js]: Agree 💯
